src/test/java/edu/harvard/iq/dataverse/util/cache/CacheFactoryBeanTest.java

```diff
@@ -0,0 +1,360 @@
+
+        // Now change the user's tier, so it is no longer limited
+        authUser.setRateLimitTier(3); // tier 3 = no limit
+        for (cnt = 0; cnt <200; cnt++) {
+            rateLimited = !cache.checkRate(authUser, action);
+            if (rateLimited) {
+                break;
+            }
+        }
+        assertTrue(!rateLimited && cnt == 200, "rateLimited:"+rateLimited + " cnt:"+cnt);
+    }
+
+    @Test
+    public void testCluster() {
+        // Make sure at least 1 entry is in the original cache
+        cache.checkRate(authUser, action);
+        String key = RateLimitUtil.generateCacheKey(authUser, action);
+
+        // Create a second cache to test cluster
+        CacheFactoryBean cache2 = new CacheFactoryBean();
+        cache2.systemConfig = mockedSystemConfig;
+        // join cluster with original Hazelcast instance
+        cache2.rateLimitCache = new TestCache(getConfig(cache.rateLimitCache.get("memberAddress")));
+
+        // Check to see if the new cache synced with the existing cache
+        assertTrue(cache.rateLimitCache.get(key).equals(cache2.rateLimitCache.get(key)));
+
+        key = "key1";
+        String value = "value1";
+        // Verify that both caches stay in sync
+        cache.rateLimitCache.put(key, value);
+        assertTrue(value.equals(cache2.rateLimitCache.get(key)));
+        // Clearing one cache also clears the other cache in the cluster
+        cache2.rateLimitCache.clear();
+        assertTrue(cache.rateLimitCache.get(key) == null);
+
+        // Verify no issue dropping one node from cluster
+        cache2.rateLimitCache.put(key, value);
+        assertTrue(value.equals(cache2.rateLimitCache.get(key)));
+        assertTrue(value.equals(cache.rateLimitCache.get(key)));
+        // Shut down hazelcast on cache2 and make sure data is still available in original cache
+        cache2.rateLimitCache.close();
+        cache2 = null;
+        assertTrue(value.equals(cache.rateLimitCache.get(key)));
+    }
+
+    private Config getConfig() {
+        return getConfig(null);
+    }
+    private Config getConfig(String members) {
+        Config config = new Config();
+        config.getNetworkConfig().getJoin().getAutoDetectionConfig().setEnabled(false);
+        config.getNetworkConfig().getJoin().getMulticastConfig().setEnabled(false);
+        config.getNetworkConfig().getJoin().getAwsConfig().setEnabled(false);
+        config.getNetworkConfig().getJoin().getAzureConfig().setEnabled(false);
+        config.getNetworkConfig().getJoin().getTcpIpConfig().setEnabled(true);
+        if (members != null) {
+            config.getNetworkConfig().getJoin().getAutoDetectionConfig().setEnabled(true);
+            config.getNetworkConfig().getJoin().getTcpIpConfig().addMember(members);
+        }
+        return config;
+    }
+
+    // convert Hazelcast IMap<String,String> to JCache Cache<Object, Object>
+    private class TestCache implements Cache<String, String>{
```

Comment:
**thought**: Wouldn't it be a lot less work to just use a cache provided by Hazelcast instead of adding a custom implementation? https://docs.hazelcast.com/hazelcast/5.3/jcache/setup

Reply:
Hazelcast only supplies an IMap and Payara supplies a JCache so I needed this to do the conversion